modules/shared/src/main/scala/org/tessellation/schema/swap.scala

```diff
@@ -0,0 +1,94 @@
+package org.tessellation.schema
+
+import org.tessellation.currency.dataApplication.DataUpdate
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.transaction.{Transaction, TransactionAmount, TransactionSalt}
+import org.tessellation.security.hash.Hash
+
+import derevo.cats.{order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import eu.timepit.refined.cats._
+import eu.timepit.refined.types.numeric.PosLong
+import io.estatico.newtype.macros.newtype
+
+object swap {
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class AllowSpendFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class CurrencyId(value: Option[Address])
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class AllowSpendReference(value: Hash) // exact type TBD (e.g. lastRef?)
+
+  @derive(decoder, encoder, order, show)
+  case class AllowSpend(
+    source: Address,
+    destination: Address,
+    currency: CurrencyId,
+    amount: TransactionAmount,
+    fee: AllowSpendFee,
+    parent: AllowSpendReference,
+    salt: TransactionSalt,
+    lastValidOrdinal: SnapshotOrdinal,
+    approverAddresses: List[Address]
+  )
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapAMMDataUpdateFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  case class PriceRange(min: TransactionAmount, max: TransactionAmount)
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class PoolId(value: Address) // exact type TBD
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SwapReference(value: Address) // exact type TBD
+
+//  @derive(decoder, encoder, order, show)
+//  @newtype
+//  case class LimitParameter(value: AnyRef) // exact type TBD
+
+  // This class likely doesn't belong here; move to appropriate location
+  @derive(decoder, encoder, order, show)
+  case class SwapAMMDataUpdate(
+    sourceAddress: Address,
+    metagraphAddress: CurrencyId,
+    swapToPair: CurrencyId,
+    fee: SwapAMMDataUpdateFee,
+    reference: SwapReference,
+    allowReference: Hash,
+    minAmount: TransactionAmount,
+    maxAmount: TransactionAmount,
+    salt: TransactionSalt,
+    maxValidGsOrdinal: SnapshotOrdinal,
+    poolId: Option[PoolId],
+    priceRange: Option[PriceRange]
+//    limitParameters: List[LimitParameter]
+  ) extends DataUpdate
+
+  @derive(decoder, encoder, order, show)
+  @newtype
+  case class SpendTransactionFee(value: PosLong)
+
+  @derive(decoder, encoder, order, show)
+  case class SpendTransaction(
+    fee: SpendTransactionFee,
+    validUntilOrdinal: SnapshotOrdinal,
+    allowSpendRef: Hash,
+    currencyId: CurrencyId,
+    amount: TransactionAmount
+  )
+
+  @derive(decoder, encoder, order, show)
+  case class AtomicAction(pending: AllowSpend, transaction: Transaction)
```

Comment:
In the current case, this would accept two SpendTransaction instances. 

The ideal implementation would be to make this a wrapper (however implemented) that could accept different transaction types so that it can be extended in the future rather than being fixed to this specific case.

Reply:
Adjusted class name, field types, and extended a trait.